lib/edge_impulse/ei_ncs.cpp

```diff
@@ -0,0 +1,404 @@
+
+	if (res == PROCESS_BUF) {
+		k_sem_give(&ei_sem);
+		res = 0;
+	}
+
+	return res;
+}
+
+int ei_ncs_clear_data(void)
+{
+	return buf_cleanup(&ei_input);
+}
+
+int ei_ncs_start_prediction(const size_t window_shift, const size_t frame_shift)
+{
+	size_t sample_shift = window_shift * ei_ncs_get_window_size() +
+			      frame_shift * ei_ncs_get_frame_size();
+
+	int ret = buf_processing_move(&ei_input, sample_shift);
+
+	if (ret == PROCESS_BUF) {
+		k_sem_give(&ei_sem);
+		ret = 0;
+	}
+
+	return ret;
+}
+
+static int raw_feature_get_data(size_t offset, size_t length, float *out_ptr)
+{
+	buf_get(&ei_input, out_ptr, offset, length);
+
+	return 0;
+}
+
+static void processing_finished(int err)
+{
+	__ASSERT_NO_MSG(user_cb);
+
+	buf_processing_end(&ei_input);
+	user_cb(err);
+}
+
+static void edge_impulse_thread_fn(void)
+{
+	signal_t features_signal;
+	features_signal.total_length = INPUT_WINDOW_SIZE;
+	features_signal.get_data = &raw_feature_get_data;
+
+	while (true) {
+		k_sem_take(&ei_sem, K_FOREVER);
+
+		/* Invoke the impulse. */
+		EI_IMPULSE_ERROR err = run_classifier(&features_signal,
+						      &ei_result, DEBUG_MODE);
+
+		if (err) {
+			LOG_ERR("run_classifier err=%d", err);
+		}
+
+		processing_finished(err);
+	}
+}
+
+static bool can_read_result(void)
+{
+	/* User is allowed to access results only from the result ready callback. */
+	return (k_current_get() == ei_thread_id);
+}
+
+int ei_ncs_get_classification_results(const char **label, float *value,
+				      float *anomaly)
+{
+	if (!can_read_result()) {
+		return -ENODATA;
+	}
+
+	float max_val = ei_result.classification[0].value;
+	size_t max_idx = 0;
+
+	for (size_t idx = 1; idx < EI_CLASSIFIER_LABEL_COUNT; idx++) {
+		if (max_val < ei_result.classification[idx].value) {
+			max_idx = idx;
+			max_val = ei_result.classification[idx].value;
+		}
+	}
+
+	*label = ei_result.classification[max_idx].label;
+	*value = ei_result.classification[max_idx].value;
+	if (EI_CLASSIFIER_HAS_ANOMALY) {
+		*anomaly = ei_result.anomaly;
+	} else {
+		*anomaly = -1.0;
+	}
+
+	return 0;
+}
+
+int ei_ncs_get_timing(int *dsp_time, int *classification_time,
+		      int *anomaly_time)
+{
+	if (!can_read_result()) {
+		return -ENODATA;
+	}
+
+	*dsp_time = ei_result.timing.dsp;
+	*classification_time = ei_result.timing.classification;
+
+	if (EI_CLASSIFIER_HAS_ANOMALY) {
+		*anomaly_time = ei_result.timing.anomaly;
+	} else {
+		*anomaly_time = -1;
+	}
+
+	return 0;
+}
+
+int ei_ncs_init(ei_ncs_result_ready_cb cb)
+{
+	static bool initialized;
+
+	if (initialized) {
+		return -EALREADY;
+	}
+
+	initialized = true;
+
+	int err = buf_cleanup(&ei_input);
+
+	__ASSERT_NO_MSG(!err);
+	ARG_UNUSED(err);
+
+	ei_thread_id = k_thread_create(&thread, thread_stack, THREAD_STACK_SIZE,
+				       (k_thread_entry_t)edge_impulse_thread_fn,
+				       NULL, NULL, NULL,
+				       THREAD_PRIORITY, 0, K_NO_WAIT);
+	k_thread_name_set(&thread, "edge_impulse_thread");
+	user_cb = cb;
```

Comment:
please set `user_cb` before thread is started
yes, in this case it does not matter much but better to keep things in right order

Reply:
done